[DAB] Add support for requirements libraries in Job Tasks

witi83 opened the pull request:
## Changes
While experimenting with DAB I discovered that requirements libraries are being ignored.

One thing worth mentioning is that `bundle validate` runs successfully, but `bundle deploy` fails. This PR only covers the second part.


## Tests
<!-- How is this tested? -->
Added a unit test

andrewnester reviewed (changes requested):
Because we could reference the requirements.txt from local file system, we need to correctly overwrite the local to remote path, something like we do for wheels
https://github.com/databricks/cli/blob/main/bundle/artifacts/artifacts.go#L181-L184

or for any jobs paths https://github.com/databricks/cli/blob/main/bundle/config/mutator/translate_paths_jobs.go#L21-L54

witi83 commented:
@andrewnester Thanks for the feedback. I'll try to work on this in the coming days.

witi83 commented:
Hey @andrewnester I updated the PR (also rebased onto latest main), please check if this goes into the right direction.

I'm not very familiar with the codebase, so I used the existing source and test code as orientation. However, I'm not sure how to write a unit test for `translate_paths_jobs`.

andrewnester commented on bundle/artifacts/artifacts.go:
```diff
@@ -186,6 +186,9 @@ func rewriteArtifactPath(b *bundle.Bundle, f *config.ArtifactFile, job *resource
 			if lib.Jar != "" && isArtifactMatchLibrary(f, lib.Jar, b) {
 				lib.Jar = remotePath
 			}
+			if lib.Requirements != "" && isArtifactMatchLibrary(f, lib.Requirements, b) {
```
You don't need to call `isArtifactMatchLibrary`, it won't work anyway because it checks that libraries defined in `artifacts` sections matches the ones used in library and it's not relevant for requirements.txt

witi83 replied:
Done, please review

andrewnester commented on bundle/artifacts/artifacts.go:
```diff
@@ -199,6 +202,9 @@ func rewriteArtifactPath(b *bundle.Bundle, f *config.ArtifactFile, job *resource
 				if lib.Jar != "" && isArtifactMatchLibrary(f, lib.Jar, b) {
 					lib.Jar = remotePath
 				}
+				if lib.Requirements != "" && isArtifactMatchLibrary(f, lib.Requirements, b) {
```
You don't need to call i`sArtifactMatchLibrary`

andrewnester commented on bundle/artifacts/artifacts_test.go:
```diff
@@ -50,6 +50,9 @@ func TestArtifactUploadForWorkspace(t *testing.T) {
 											Whl: "/Workspace/Users/foo@bar.com/mywheel.whl",
 										},
+										{
+											Requirements: "/Workspace/Users/foo@bar.com/requirements.txt",
```
This does not really test anything useful. It should instead point to local requirements.txt and then later assert that this path is translated correctly to remote path

andrewnester replied:
Having such test would catch an error below where local to remote translation is not happening

andrewnester commented on bundle/config/mutator/translate_paths_jobs.go:
```diff
@@ -50,6 +50,11 @@ func rewritePatterns(t *translateContext, base dyn.Pattern) []jobRewritePattern
 		},
+		{
+			base.Append(dyn.Key("libraries"), dyn.AnyIndex(), dyn.Key("requirements")),
+			t.translateNoOp,
```
You have to do the translation here and not skip it

witi83 replied:
Done, please review

witi83 commented:
@andrewnester Thanks for the extensive feedback. I adapted the PR accordingly. I was also able to add a test in `translate_paths_test`

andrewnester commented on bundle/config/artifact.go:
```diff
@@ -19,6 +19,7 @@ func (artifacts Artifacts) ConfigureConfigFilePath() {
 type ArtifactType string
 
 const ArtifactPythonWheel ArtifactType = `whl`
+const ArtifactPythonRequirementsFile ArtifactType = `requirements`
```
requirements is not an artifact so we don't need to a new type but instead we should treat it as normal file

andrewnester commented on bundle/artifacts/artifacts_test.go:
```diff
@@ -84,26 +100,39 @@ func TestArtifactUploadForWorkspace(t *testing.T) {
+		filer.CreateParentDirectories,
+	).Return(nil)
+
+	requirementsError := uploadArtifact(context.Background(), b, b.Config.Artifacts["requirements"], "/foo/bar/artifacts", requirementsMockFiler)
```
I think this test is passing only because you added the following section in the test but that's not something users will configure

```
"requirements": {
					Type: config.ArtifactPythonRequirementsFile,
					Files: []config.ArtifactFile{
						{Source: requirementsLocalPath},
					},
				},
```

andrewnester reviewed (changes requested):
Thanks for the contribution! I think there's a bit a confusion about approach we need to take, I guess it's better if I explain expectations a bit more.

The way this should work is customers specify the following section in their bundle config

```
libraries:
- requirements: ./local/path/to/requirements.txt
```

and nothing else.

What DABs should do is
1. Upload requirements.txt to workspace file system - this is done automatically similarly to notebooks for example
2. Rewrite local path to remote path (where requirements.txt was uploaded) in bundle config similarly to how this happens for notebooks now. - This is what needs to be implemented in this PR.

You don't need to use artifacts because artifacts is something that is being built, packaged and then uploaded by DABs, for requirements.txt is not needed.

I hope this clarifies what needs to be done

andrewnester commented on bundle/artifacts/artifacts_test.go:
```diff
@@ -23,6 +23,10 @@ func TestArtifactUploadForWorkspace(t *testing.T) {
 	testutil.Touch(t, whlFolder, "source.whl")
 	whlLocalPath := filepath.Join(whlFolder, "source.whl")
 
+	requirementsFolder := filepath.Join(tmpDir, "requirements")
```
I don't think you need to modify this test because requirements.txt is not related to artifacts. Having a translate_path test you added is sufficient

andrewnester commented on bundle/artifacts/artifacts.go:
```diff
@@ -186,6 +186,9 @@ func rewriteArtifactPath(b *bundle.Bundle, f *config.ArtifactFile, job *resource
 			if lib.Jar != "" && isArtifactMatchLibrary(f, lib.Jar, b) {
 				lib.Jar = remotePath
 			}
+			if lib.Requirements != "" {
```
Now with having translate_path change for requirements I don't think you need this

andrewnester commented on bundle/artifacts/artifacts.go:
```diff
@@ -199,6 +202,9 @@ func rewriteArtifactPath(b *bundle.Bundle, f *config.ArtifactFile, job *resource
 				if lib.Jar != "" && isArtifactMatchLibrary(f, lib.Jar, b) {
 					lib.Jar = remotePath
 				}
+				if lib.Requirements != "" {
```
Same as above

witi83 commented:
Hey @andrewnester

Thank you very much! As mentioned, as I'm new to this codebase, I really appreciate your guidance! It's definitely a quite interesting task to get familiar with DAB. 😉

I reverted the changes regarding artifacts (as a requirements.txt is not seen as an artifact).

Please check if it now makes sense to you

andrewnester commented:
Confirmed manually that requirements.txt is uploaded and used correctly on DBR 15.1+

andrewnester commented:
Thanks for contribution @witi83 !
